XML config files for Live Broadcasting preferences

Palakis opened the pull request:
These contributions add an XML settings system for Live Broadcasting settings which allows management of several named sets of settings (a.k.a "profiles").

This is phase 1 out of 2 of my [GSoC project for Mixxx](https://www.mixxx.org/wiki/doku.php/multi-broadcasting) to add multi-broadcasting to Mixxx (being able to stream to several servers simultaneously). Each broadcasting output will have its own profile

This PR aims to implement this subsystem into Mixxx while keep the current Live Broadcasting preferences UI, which will be refactored in phase 2. Settings are saved in an XML document named `Default Profile.bcp.xml` located in the `broadcast_profiles` subfolder of Mixxx's settings folder.

__Tasks:__
- Implement the BroadcastProfile class:
  - [X] Basic structure
  - [X] XML read/write code
  - [x] Validate proper functioning
  - [x] Write unit tests
- Refactor the BroadcastSettings class
  - [x] Make sure only one instance of BroadcastSettings exists
  - [x] QList of BroadcastProfile instances
  - [x] List profiles from filesystem
- [x] Automatic upgrade from legacy Live Broadcasting preferences format (in mixxx.cfg) to new XML engine

Commit:
preferences: add WIP class for broadcasting profiles

Commit:
xml profile schema: refactor tag names to SnakeCase

Commit:
Fix errors in BroadcastProfile and make sure it builds

I initially forgot to add broadcastprofile.cpp into the list
of source files in build/depends.py, so I didn't noticed all the
errors and mistakes I've made while writing this earlier.
This commit aims to fix these errors and have the BroadcastProfile
class to build.

Commit:
(WIP) BroadcastSettings and BroadcastProfile refactor

No idea if this works. I just want to make sure it is saved
somewhere out of my computer.
I may squash this with other commits later.

Commit:
BroadcastSettings: fix behaviour of filename gen. method

Commit:
BroadcastProfile: fix XML load/save code

Both ways didn't handled the document's root tag.
Now, profile values are saved properly.

Commit:
BroadcastSettings: enhance default profile creation

Commit:
BroadcastSettings: add qDebug statements

Commit:
BroadcastProfile: remove extra qualifications in header file

Commit:
Merge branch 'master' of https://github.com/mixxxdj/mixxx into broadcast-profiles

Commit:
BroadcastProfile: fix XML file load code

Commit:
BroadcastSettings: make sure only one instance exists

daschuer commented on src/broadcast/broadcastmanager.cpp:
```diff
@@ -6,14 +6,15 @@
 #include "soundio/soundmanager.h"
 
-BroadcastManager::BroadcastManager(UserSettingsPointer pConfig,
+BroadcastManager::BroadcastManager(SettingsManagerPointer pSettingsManager,
```
since you do not store the pointer, you can pass it as plain pointer.

Palakis replied:
@daschuer Same when passing it to DlgPreferences?

daschuer replied:
Yes. Shard pointer for shared ownership. With a plain pointer parameter, the function promises not to store the pointer.

Palakis replied:
@daschuer Okay. If thinking this way, I think I should use shared pointer for `BroadcastProfile` instances.

daschuer commented on src/encoder/encoderbroadcastsettings.cpp:
```diff
@@ -10,9 +10,9 @@
-{
+EncoderBroadcastSettings::EncoderBroadcastSettings(
+        BroadcastSettingsPointer settings) :
+m_settings(settings) {
```
Style nit: 
```C++
EncoderBroadcastSettings::EncoderBroadcastSettings(
        BroadcastSettingsPointer settings) 
        : m_settings(settings) {
```

Commit:
Code style fix in EncoderBroadcastSettings

Commit:
SettingsManager: remove shared pointer type

This isn't needed in the current state of the code.

Commit:
Add auto migration from legacy broadcasting settings

These additions automatically fetches existing
Live Broadcasting preferences from mixxx.cfg, puts
them in the Default Profile and removes the now
useless keys from mixxx.cfg

Commit:
Broadcasting settings: don't remove previous settings when migrating

Commit:
Move BroadcastProfile::loadLegacySettings to BroadcastSettings

Commit:
Unindent anonymous namespace of BroadcastProfile class

daschuer commented:
Nice, please give a hint when this is worth to test manually.

Commit:
BroadcastProfile: remove non-existent method declaration

Commit:
Add unit tests for BroadcastProfile class

Commit:
Fix and rename method BroadcastProfile::checkNameCompliance

Commit:
BroadcastProfile: fix DefaultValues unit test

Commit:
BroadcastProfile: fix typo in forbidden characters regex

Add more tests in the process, because the regex changed.

sblaisot commented:
probably you need to take a look at upgrade.cpp/.h.
This is where migrating to old preference style to new ones should reside.

Commit:
BroadcastProfile: fix forbidden chars regex syntax

Palakis commented:
@daschuer I believe it is. Gonna do testing from my end.

daschuer commented on src/preferences/broadcastprofile.cpp:
```diff
@@ -0,0 +1,466 @@
+    return profile;
+}
+
+void BroadcastProfile::setDefaultValues() {
```
this function does not set the default values. It resets the profile back to default. 
How about "adoptDefaultValues()"?

daschuer commented on src/preferences/broadcastprofile.cpp:
```diff
@@ -0,0 +1,466 @@
+}
+
+void BroadcastProfile::loadValues(const QString& filename) {
+    QDomElement doc = XmlParse::openXMLFile(filename, kDoctype);
```
What happens if this call fails?

Palakis replied:
A call to `XmlParse::openXMLFile` on a non-existing document pops up an error message ("There was a problem parsing the XML file [...]") and returns an empty `QDomElement`. In my current codebase I added a children count check after that call to avoid writing to attributes if the element has no children.

daschuer commented on src/preferences/broadcastprofile.cpp:
```diff
@@ -0,0 +1,466 @@
+    doc.appendChild(docRoot);
+
+    QFile xmlFile(filename);
+    xmlFile.open(QIODevice::WriteOnly | QIODevice::Text);
```
Error handling?

Palakis replied:
Woops. Fixed. Now this method and `loadValues` have error handling and return a bool to signal the success or failure of the call.

daschuer commented on src/preferences/broadcastprofile.cpp:
```diff
@@ -0,0 +1,466 @@
+const QString kDefaultStreamGenre = QObject::tr("Live Mix");
+const bool kDefaultStreamPublic = false;
+
+const QRegExp kForbiddenChars = QRegExp("[<>:\"\\/|?*\\\\]");
```
I am not sure if this is complete. I have just found this file: 
https://github.com/martamius/mushi/blob/master/qtMushi/utils/filenamevalidatinglineedit.cpp
Maybe we can re use it in Mixxx.

Palakis replied:
I believe this can all be done in the regex, so I have to update it to account for Windows device names and the `..` substring.

daschuer commented on src/preferences/broadcastprofile.cpp:
```diff
@@ -0,0 +1,466 @@
+    // Replace occurences of forbidden characters with a space
+    // to avoid problems with the underlying filesystem.
+    // Char list comes from MSDN article "Naming Files, Paths, and Namespaces".
+    m_profileName.replace(kForbiddenChars, " ");
```
We should either store the original name inside the file, or ask the user to pick a valid name.

Palakis replied:
The latter is what I planned for the UI and the reason why there's that `BroadcastProfile::validName` method. However you're right for the former. I need to take some time to think about this in regards to that other comment on the profiles list/hashmap in `BroadcastSettings`.

daschuer commented on src/preferences/broadcastprofile.cpp:
```diff
@@ -0,0 +1,466 @@
+}
+
+bool BroadcastProfile::getEnabled() const {
+    return m_enabled;
```
What mean enabled actually? Please add a comment.

Palakis replied:
It was already there in the previous BroadcastSettings, so I kept it because it might come in handy for multi-broadcasting (this would mean: "this connection is enabled").

daschuer replied:
Ok, please add a clarifying comment to the source.

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
-            ConfigKey(kConfigKey, kMaximumRetries), getDefaultMaximumRetries());
+    : m_pConfig(pConfig),
+    m_profiles(),
+    m_currentProfile(kDefaultProfile) {
```
Is this prepared for n profiles? If we merge this early, I would like to avoid a future migration for n profiles.

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
-        return getDefaultPort();
+BroadcastSettings::~BroadcastSettings() {
+    for(BroadcastProfile* profile : m_profiles) {
+        delete profile;
```
we should try to get rid of this delete by using smart pointers or such.

Palakis replied:
I believe `QSharedPointer`s would be a good fit. What do you think?

daschuer replied:
In case we have a shared ownership later yes. 
If m_profiles is the only owner it should be a std::unique_ptr container.

Palakis replied:
Currently I rewrote this with `QSharedPointer`s because I had trouble when passing references to `std::unique_ptr`s. I'll have to come back to this later.

Palakis replied:
Managed to fix this in the latest commits!

daschuer commented on src/preferences/broadcastsettings.h:
```diff
@@ -1,97 +1,39 @@
+
     // Pointer to config object
     UserSettingsPointer m_pConfig;
+    QList<BroadcastProfile*> m_profiles;
```
I think it is best to replace it with 
std::vector<std::unique_prt<BroadcastProfile> 
like in searchquerry.h

daschuer replied:
or std::hash

Palakis replied:
Fixed, I replaced the `QList<BroadcastProfile*` with an `std::vector<std::unique_ptr<BroadcastProfile>>`.

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
+void BroadcastSettings::loadProfiles() {
+    QDir profilesFolder(getProfilesFolder());
+    if(!profilesFolder.exists()) {
+        kLogger.warning()
```
this is not a warning, since it is normal, right?

daschuer replied:
use debug() for such messages.

Palakis replied:
Woops, sorry. Gonna fix that.

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
+    QFileInfoList files =
+            profilesFolder.entryInfoList(nameFilters, QDir::Files, QDir::Name);
+
+    if(files.count() > 0) {
```
This requires some comments, about the migration strategy

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
+        kLogger.warning()
+                << "No profiles found. Creating default profile.";
+
+        // Create default profile
```
useless comment.

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
-    return m_pConfig->getValue(
-            ConfigKey(kConfigKey, kStreamGenre),
-            getDefaultStreamGenre());
+QString BroadcastSettings::filenameForProfile(const QString& profileName) {
```
FilePathForProfile

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
-            getDefaultStreamName());
+QString BroadcastSettings::getProfilesFolder() {
+    QString profilesPath(m_pConfig->getSettingsPath());
+    profilesPath.append(QDir::separator() + QString(kProfilesSubfolder));
```
Qt is very smart in handling stings. 
profilesPath.append(QDir::separator() + kProfilesSubfolder);

Palakis replied:
Doing so causes errors in the compiler: `operator +: 4 overloads have similar conversions `

Palakis replied:
Precisely (according to the lines following that error):
- `QTransform operator +(const QTransform &,qreal)`
- `const QSizeF operator +(const QSizeF &,const QSizeF &)`
- `const QSize operator +(const QSize &,const QSize &)`
- `const QPointF operator +(const QPointF &,const QPointF &)`
- `const QPoint operator +(const QPoint &,const QPoint &)`
- `const QString operator +(const QString &,const QByteArray &)`
- `const QString operator +(const QByteArray &,const QString &)`
- `const QString operator +(const QString &,char)`
- `const QString operator +(char,const QString &)`
- `const QString operator +(const char *,const QString &)`
- `const QString operator +(const QString &,const char *)`
- `const QString operator +(QChar,const QString &)`
- `const QString operator +(const QString &,QChar)`
- `const QString operator +(const QString &,const QString &)`
- `const QByteArray operator +(char,const QByteArray &)`
- `const QByteArray operator +(const char *,const QByteArray &)`
- `const QByteArray operator +(const QByteArray &,char)`
- `const QByteArray operator +(const QByteArray &,const char *)`
- `const QByteArray operator +(const QByteArray &,const QByteArray &)'

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
-    m_pConfig->setValue(ConfigKey(kConfigKey, kStreamName), value);
-}
+void BroadcastSettings::setCurrentProfile(BroadcastProfile *profile) {
+    if(!profile)
```
do we need this for clear the current profile?

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
-void BroadcastSettings::setStreamName(const QString& value) {
-    m_pConfig->setValue(ConfigKey(kConfigKey, kStreamName), value);
-}
+void BroadcastSettings::setCurrentProfile(BroadcastProfile *profile) {
```
I think this function is not n-profile proofed.

daschuer commented on src/preferences/broadcastsettings.cpp:
```diff
@@ -1,417 +1,135 @@
-void BroadcastSettings::setStreamPublic(bool value) {
-    m_pConfig->setValue(ConfigKey(kConfigKey, kStreamPublic), value);
+BroadcastProfile* BroadcastSettings::getProfileByName(const QString& profileName) {
+    for(BroadcastProfile* profile : m_profiles) {
```
or m_profiles should be a std::hash for fast profile name lookup.

Palakis replied:
I just noticed that the profile name attribute in `BroadcastProfile` is useless in it (not used in load/save code and the XML document itself), so maybe name "keeping" should be the responsibility of `BroadcastSettings`?
That's why I initially used a list instead of a hashmap, to avoid having to synchronize names between the hashmap and the instances.

daschuer replied:
Ah OK, so if someone changes the name, the list does not care. 
You should place a comment about it.

Palakis replied:
I refactored `BroadcastSettings::m_profiles` as a `std::map<QString, BroadcastProfilePtr>`. The map key and the profile name are synchronized with a signal/slot pair triggered when a call to `BroadcastProfile::setProfileName` is made.
What do you think of this solution?

daschuer commented:
This looks already very good, I have left some minor comments. 

I think if we commit this to master now, we loose some degrees of freedom. During the later project, so I think we should create a "stable" project branch, where we merge this and all following iterations too. 
What do you think?

Palakis commented:
@daschuer Good idea. Will create this on my repo, certainly under the name `gsoc-stable`, and merge my work into it once all comments are sorted out.

daschuer commented:
> @daschuer Good idea. Will create this on my repo, certainly under the name gsoc-stable, and merge my work into it once all comments are sorted out.

Fine, and do a Pull request for this to have it visible on the mixxxdj/mixxx repro.

Commit:
Several minor fixes in BroadcastProfile class

Commit:
Use a QSharedPointer smart pointer for instances of BroadcastProfile

Commit:
Several changes in BroadcastSettings class

Commit:
Minor fixes in DlgPrefBroadcast

Commit:
BroadcastProfile: replace use of QSharedPointer with std::unique_ptr

Commit:
BroadcastSettings: hashmap for fast lookup of profiles by name

daschuer commented on src/preferences/broadcastprofile.cpp:
```diff
@@ -0,0 +1,487 @@
+        return nullptr;
+
+    QString profileName = xmlFile.baseName();
+    BroadcastProfile* profile = new BroadcastProfile(profileName);
```
here you should use std::make_unique<> and return the BroadcastProfilePtr

Palakis replied:
Fixed!

daschuer commented on src/test/broadcastprofile_test.cpp:
```diff
@@ -0,0 +1,204 @@
+
+    // Load XML file using static loadFromFile and assert
+    // the discriminating value is present
+    BroadcastProfilePtr savedProfile = BroadcastProfile::loadFromFile(filename);
```
This does not build, but will be fixed once you return a BroadcastProfilePtr.

Palakis replied:
Woops. Gonna fix this ASAP.

Commit:
BroadcastProfile: loadFromFile returns a smart pointer

daschuer commented:
> I refactored BroadcastSettings::m_profiles as a std::map<QString,
BroadcastProfilePtr>. The map key and the profile name are synchronized
with a signal/slot pair triggered when a call to
BroadcastProfile::setProfileName is made.
What do you think of this solution?


Perfect!

Commit:
Remove uses of C++14 std::make_unique

Commit:
Revert removal of std::make_unique and use it properly

Palakis commented:
I forgot to connect `BroadcastProfile::profileNameChanged` to `BroadcastSettings::onProfileNameChanged`. Since `std::unique_ptr` is not compatible with Qt's object system, I replaced it with `QSharedPointer`, as it is easy to pass around.

Commit:
Use QSharedPointers instead of std::unique_ptr and add missing connect call

BroadcastProfile notifies name changes to BroadcastSettings with Qt's signals &
slots system. However Qt is not compatible with the std::unique_ptr smart
pointer. A refactor of BroadcastProfilePtr is required to make it compatible.
std::unique_ptr has been replaced with QSharedPointer, as this type is easy
to pass around in the code while keeping a reference to the original pointer
and deleting it when not used anymore.

Commit:
Refactor BroadcastProfile ForbiddenChars unit test

daschuer commented:
[  FAILED  ] BroadcastProfileTest.ForbiddenChars

Else it works nice. Thank you.

Palakis commented:
@daschuer Did you tried with the latest changes? This assert call was removed in one of them.

daschuer commented:
I have rebuid and tested it again, and it still fails. Is it a linux only issue?

Commit:
BroadcastProfile tests: remove deprecated asserts

Palakis commented:
My bad, I was supposed to remove `validProfile` and `invalidProfile` in the previous commit, since those are deprecated because the profile name isn't stripped of forbidden chars anymore.

Commit:
BroadcastProfile: add clarifying comment

daschuer commented:
This looks goo to me so far, is there anything left to do in this branch?

Palakis commented:
@daschuer Not much left to do: making sure it is ready for handling several broadcasting outputs/profiles.
One way I can come up with is to remove `setCurrentProfile` and `getCurrentProfile` (which won't be of use for multi-broadcasting), turn BroadcastSettings into a Qt read-only model and implement an overload of `QAbstractItemModel::data` that directly returns a `BroadcastProfilePtr` instead of a `QVariant`. This way I can get profiles by index and replace calls to `getCurrentProfile` with `data`.

Commit:
BroadcastSettings: implement model and remove get/setCurrentProfile

Commit:
BroadcastProfile unit testing: use regular tests

Commit:
BroadcastProfile tests: remove useless fixture

Palakis commented:
I believe there's nothing left to do on this. As usual, I'm open for comments on these contributions!
More additions are needed in `BroadcastSettings` for multi-broadcasting (e.g. : a method to rename a profile), but I believe those will come during phase 2 of GSoC. I'll also work on unit tests for `BroadcastSettings`.

daschuer commented:
Fine, am happy with this as well so you can merge it to your GSoC branch and advance to the next PR. 
Thank you for the work so far. :-)

Be-ing commented:
Should this PR be closed?

Palakis commented:
Yes, absolutely. Work continues in #1300.
